java/arcs/core/common/CounterFlow.kt

```diff
@@ -0,0 +1,33 @@
+package arcs.core.common
+
+import kotlinx.atomicfu.atomic
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+
+/**
+ * Provides a counter with increment/decrement functionality whose value can be observed via a
+ * Flow<Int>.
+ *
+ * The counter value will be conflated: if a large number of changes occur at once, collectors of
+ * the flow may not receive all of them, but are guaranteed to receive the most recent one.
+ */
+class CounterFlow(initialValue: Int = 0) {
+    private val count = atomic(initialValue)
+    private val stateFlow = MutableStateFlow<Int>(initialValue)
+
+    /**
+     * Returns a [Flow<Int>] that emits counter changes. Not every change is guaranteed to be emitted, but
+     * you're always guaranteed to receive the last one.
+     */
+    val flow: Flow<Int> = stateFlow
+
+    /**
+     * Increments the counter. The change will be emitted on any active collections of [flow].
+     */
+    fun increment() { stateFlow.value = count.incrementAndGet() }
```

Comment:
Feel there is a bug: unexpected `stateFlow.value` when there are racing `increment()` or `decrement()` or both.
The atomic `count` works as it is given two racing `increment()` i.e. 0->1 and 1->2.
The assignment `stateFlow.value =` could be still racing, so its value could be 1 or 2.

Though it is not causing any issues given serialized api calls.
The comment is for the `CounterFlow` impl itself.

Reply:
Yes you're right :(